src/optimizer/query_to_operator_transformer.cpp

```diff
@@ -0,0 +1,644 @@
+    auto insert_expr =
+        new OperatorExpression(LogicalInsertSelect::Make(target_db_id, target_ns_id, target_table_id), {});
+    op->GetSelect()->Accept(this, parse_result);
+    insert_expr->PushChild(output_expr_);
+    output_expr_ = insert_expr;
+    return;
+  }
+
+  // column_objects represents the columns for the current table as defined in its schema
+  auto column_objects = accessor_->GetSchema(target_table_id).GetColumns();
+
+  // vector of column oids
+  std::vector<catalog::col_oid_t> col_ids;
+
+  // INSERT INTO table_name VALUES (val1, val2, ...), (val_a, val_b, ...), ...
+  if (op->GetInsertColumns()->empty()) {
+    for (const auto &values : *(op->GetValues())) {
+      if (values.size() > column_objects.size()) {
+        throw CATALOG_EXCEPTION("ERROR:  INSERT has more expressions than target columns");
+      }
+      if (values.size() < column_objects.size()) {
+        for (auto i = values.size(); i != column_objects.size(); ++i) {
+          // check whether null values or default values can be used in the rest of the columns
+          if (!column_objects[i].Nullable() && column_objects[i].StoredExpression() == nullptr) {
+            throw CATALOG_EXCEPTION(
+                ("ERROR:  null value in column \"" + column_objects[i].Name() + "\" violates not-null constraint")
+                    .c_str());
+          }
+        }
+      }
+    }
+    for (const auto &col : column_objects) col_ids.push_back(col.Oid());
+
+  } else {
+    // INSERT INTO table_name (col1, col2, ...) VALUES (val1, val2, ...), ...
+    auto num_columns = op->GetInsertColumns()->size();
+    for (const auto &tuple : *(op->GetValues())) {  // check size of each tuple
+      if (tuple.size() > num_columns) {
+        throw CATALOG_EXCEPTION("ERROR:  INSERT has more expressions than target columns");
+      }
+      if (tuple.size() < num_columns) {
+        throw CATALOG_EXCEPTION("ERROR:  INSERT has more target columns than expressions");
+      }
+    }
+
+    // set below contains names of columns mentioned in the insert statement
+    std::unordered_set<catalog::col_oid_t> specified;
+    auto schema = accessor_->GetSchema(target_table_id);
+
+    for (const auto &col : *(op->GetInsertColumns())) {
+      try {
+        const auto &column_object = schema.GetColumn(col);
+        specified.insert(column_object.Oid());
+      } catch (const std::out_of_range &oor) {
+        throw CATALOG_EXCEPTION(
+            ("ERROR:  column \"" + col + "\" of relation \"" + target_table->GetTableName() + "\" does not exist")
+                .c_str());
+      }
+    }
+
+    for (const auto &column : schema.GetColumns()) {
+      // this loop checks not null constraint for unspecified columns
+      if (specified.find(column.Oid()) == specified.end() && !column.Nullable() &&
+          column.StoredExpression() == nullptr) {
+        // TODO(peloton): Add check for default value's existence for the current column
```

Comment:
If `column.StoredExpression()` is **not** null, then does that mean that there is a default expression that we can use for this column for the INSERT?

If yes, then I think we can get rid of this TODO. Otherwise, I don't understand what this TODO is asking for.